In [1]:
import pandas as pd
import numpy as np
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.


In [2]:
df = np.loadtxt("no_header.csv", delimiter = ",")
print("Initial dataset is dimention: ", df.shape)

np.random.shuffle(df)

X = df[:, :-1]
Y = df[:, -1]

modalities = int(max(Y)) # knowing the 

Initial dataset is dimention:  (322, 103)


We want to create an output vector Y of dimention 322, like our instances, and 3, like the number of modalities it can assume. We put 1 in correspondence of the modality the vector assumes. 

So, for example, if an istance belongs to group 2, the corresponding Y vector will be: [0, 1, 0].

In [3]:
y_ = [0 for i in range(modalities)]
Y_ = []

for i in Y:
    y_[int(i)-1] = 1
    Y_.append(y_)
    y_ = [0 for i in range(modalities)]


Y = np.array(Y_)
Y.shape

(322, 3)

We now split our dataset into two sets: one training set and a test set.

In [4]:

test_bound = int(round(df.shape[0]*0.9))
X_train = X[:test_bound-1, :]
Y_train = Y[:test_bound-1, :]

X_test = X[test_bound:, :]
Y_test = Y[test_bound:, :]

We build and compile our 3 layers neural network. The input will be as long as the attributes of the X, the output will be as long as the modalities of our vector Y.

In [5]:
model = Sequential()
model.add(Dense(10, input_dim = X.shape[1], activation = "relu"))
model.add(Dense(15, activation = "relu"))
model.add(Dense(modalities, activation = "sigmoid"))

model.compile(optimizer = "adam", loss="binary_crossentropy", metrics=["accuracy"])

In [6]:
# let's train our model
model.fit(X_train,
         Y_train,
         epochs = 300,
         batch_size = 10,
         verbose = False)

In [8]:
predictions = model.predict(X_test)
# rounding our predictions
rounded = []
for i in predictions:
    rounded.append([round(x) for x in i])
rounded = np.array(rounded)

# let's evaluate our predictions
print("Model efficiency: %s per cent" %(model.evaluate(X_test, Y_test)[1]*100))
print(rounded == Y_test)

32/32 [==============================] - 0s
Model efficiency: 95.8333313465 per cent
[[ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [False False  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [False False  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [ True  True  True]]
